src/Microsoft.OData.Core/Metadata/EdmLibraryExtensions.cs

```diff
@@ -595,13 +595,26 @@ internal static bool IsPrimitiveType(Type clrType)
         {
             Debug.Assert(clrType != null, "clrType != null");
 
-            if (clrType == typeof(UInt16) || clrType == typeof(UInt32) || clrType == typeof(UInt64))
-            {
+            // Directly check the most common types for first
+            // for better performance.
+            return clrType == typeof(string)

```

Comment:
Thanks @chrisspre. I'm running your suggestion as well as other variations I have thought about as a result of your suggestion under a benchmark. I'll report back with my findings.

Reply:
I test the following approaches:
- using a pre-initialized `HashSet<TypeCode>` lookup first then the fallback to the dictionary when the type code is not in the hashset: `return typeCodeSet.Contains(typeCode) || primitiveMap.ContainsKey(type) ...`
- start with a lookup in a pre-initialized `bool[]` array where the type code is the index then fallback to the dictionary if the typecode is not in the array: `return typeCodeArray[typeCode] || primitiveMap.ContainsKey(type) ...`
- use an int bitmap instead of a bool array, where a bit is set to one when its position in the int matches a supported type code: `return typeCodeBitMap & (1 << typeCode) == (1 << typeCode) || primitiveMap.ContainsKey(type) ...`

See full implementations here: https://github.com/habbes/odata.net/blob/entity-type-perf-tests/src/Microsoft.OData.Core/Metadata/EdmLibraryExtensions.cs

Find full benchmarks here: https://github.com/habbes/odata.net/blob/entity-type-perf-tests/test/PerformanceTests/ComponentTests/CoreLib/PrimitiveValueTests.cs

I found that the approach using `HashSet<TypeCode>` slower than my initial PR ("linear scan" using a chain of type comparisons) for the most common types (string, int, bool, Guid, DateTimeOffset), but slightly faster for the lesser common types (the ones looked up in the dictionary.

I found that the using bool array and int bitmap were much faster than using the `HashSet` on all counts. The array and int bitmap approaches had similar performances. For the bool array and int map approaches, I also found that it may be beneficial to add direct checks for DateTimeOffset (and Guid) which are common types for which don't have dedicated TypeCodes. It drastically speeds checks for DateTimeOffset and Guid while slightly slowing down types in the dictionary or spatial types.

The bool array and int bitmap are slightly slower than the original PR (using direct comparison chains) for strings and ints.

| Method                              | String    | Int       | Bool      | DateTimeOffset      | Guid      | ByteArray | NullableInt | Uint64    | TimeOfDay | Point     |
|-------------------------------------|-----------|-----------|-----------|-----------|-----------|-----------|-------------|-----------|-----------|-----------|
| Baseline (existing impl)                     | 23.318 ns | 22.829 ns | 22.224 ns | 22.544 ns | 22.688 ns | 24.875 ns | 78.282 ns    | 6.486 ns  | 22.577 ns | 34.164 ns |
| Original PR (chain of conditions)          | 4.205 ns  | 5.338 ns  | 6.994 ns  | 11.713 ns | 11.734 ns | 42.085 ns | 53.779 ns    | 21.584 ns | 44.069 ns | 51.081 ns |
| HashSet<TypeCode>          | 17.417 ns | 16.077 ns | 17.013 ns | 15.116 ns | 18.298 ns | 36.603 ns | 64.565 ns    | 16.758 ns | 35.429 ns | 48.322 ns |
| bool[] map          | 6.454 ns  | 6.377 ns  | 6.846 ns  | 20.962 ns | 21.927 ns | 22.911 ns | 56.591 ns    | 6.085 ns  | 22.108 ns | 32.827 ns |
| bool[] map with checks for DateTimeOffset and Guid          | 6.832 ns  | 6.477 ns  | 6.25 ns   | 7.846 ns  | 10.712 ns | 26.788 ns | 55.688 ns    | 6.252 ns  | 28.462 ns | 40.168 ns |
| int bitmap          | 6.203 ns  | 6.015 ns  | 6.013 ns  | 20.36 ns  | 24.865 ns | 22.623 ns | 61.293 ns    | 7.822 ns  | 20.267 ns | 36.182 ns |
| int bitmap with checks for DateTimeOffset and Guid           | 8.715 ns  | 8.236 ns  | 6.675 ns  | 8.772 ns  | 9.063 ns  | 28.058 ns | 51.053 ns    | 6.217 ns  | 26.331 ns | 38.813 ns |

Reply:
Based on these results, I've settled for the int bitmap approach with extra checks for Guid and DateTimeOffset. I think it has reasonable trade-offs. It could also consider the array approach for arguably better readability, but I thought we don't need that extra allocation.